lib/pdk/module/release.rb

```diff
@@ -0,0 +1,260 @@
+          raise PDK::CLI::ExitWithError, _('The module is not Forge compatible') if requires_forge_compatibility? && !forge_compatible?
+        end
+
+        # Note that these checks are duplicated in the run_publish method, however it's a much better
+        # experience to fail early, than going through the whole process, only to error at the end knowing full well
+        # it'll fail anyway.
+        validate_publish_options!
+
+        run_validations(options) unless skip_validation?
+
+        PDK.logger.info _('Releasing %{module_name} - from version %{module_version}') % {
+          module_name:    module_metadata.data['name'],
+          module_version: module_metadata.data['version'],
+        }
+
+        PDK::Util::ChangelogGenerator.generate_changelog unless skip_changelog?
+
+        # Calculate the new module version
+        new_version = specified_version
+        if new_version.nil? && !skip_changelog?
+          new_version = PDK::Util::ChangelogGenerator.compute_next_version(module_metadata.data['version'])
+        end
+        new_version = module_metadata.data['version'] if new_version.nil?
+
+        if new_version != module_metadata.data['version']
+          PDK.logger.info _('Updating version to %{module_version}') % {
+            module_version: new_version,
+          }
+
+          # Set the new version in metadata file
+          module_metadata.data['version'] = new_version
+          write_module_metadata!
+
+          # Update the changelog with the correct version
+          PDK::Util::ChangelogGenerator.generate_changelog unless skip_changelog?
+        end
+
+        run_documentation(options) unless skip_documentation?
+
+        run_dependency_checker(options) unless skip_dependency?
+
+        if skip_build?
+          # Even if we're skipping the build, we still need the name of the tarball
+          # Use the specified package path if set
+          package_file = specified_package if package_file.nil?
+          # Use the default as a last resort
+          package_file = default_package_filename if package_file.nil?
+        else
+          package_file = run_build(options)
+        end
+
+        run_publish(options.dup, package_file) unless skip_publish?
+      end
+
+      def module_metadata
+        @module_metada ||= PDK::Module::Metadata.from_file(File.join(module_path, 'metadata.json'))
+      end
+
+      def write_module_metadata!
+        module_metadata.write!(File.join(module_path, 'metadata.json'))
+        clear_cached_data
+      end
+
+      def default_package_filename
+        return @default_tarball_filename unless @default_tarball_filename.nil?
+        builder = PDK::Module::Build.new(module_dir: module_path)
+        @default_tarball_filename = builder.package_file
+      end
+
+      def run_validations(opts)
+        # TODO: Surely I can use a pre-existing class for this?
```

Comment:
I assume the plan is to come back and dedupe this & the `pdk validate` CLI in a follow up PR?

Reply:
Hasn't been ticketted yet. Need to do that first.  Seemed out of scope for this PR.

Reply:
Raised https://tickets.puppetlabs.com/browse/PDK-1552